pkg/virt-api/rest/subresource.go

```diff
@@ -517,6 +518,98 @@ func (app *SubresourceAPIApp) RestartVMRequestHandler(request *restful.Request,
 	response.WriteHeader(http.StatusAccepted)
 }
 
+func (app *SubresourceAPIApp) RenameVMRequestHandler(request *restful.Request, response *restful.Response) {
+	name := request.PathParameter("name")
+	namespace := request.PathParameter("namespace")
+
+	if request.Request.Body == nil {
+		writeError(errors.NewBadRequest("Request with no body, a new name is expected as the request body"),
+			response)
+		return
+	}
+
+	// Read the new name of the VM
+	newNameBuf, err := ioutil.ReadAll(request.Request.Body)
+
+	if err != nil {
+		writeError(errors.NewConflict(v1.Resource("virtualmachine"),
+			name, err), response)
+		return
+	}
+
+	newName := string(newNameBuf)
+
+	request.Request.Body.Close()
+
+	if newName == "" {
+		writeError(errors.NewBadRequest("The new name is empty"), response)
+		return
+	}
+
+	if name == newName {
+		writeError(errors.NewBadRequest("The new name cannot be identical to the old name"), response)
+		return
+	}
+
+	// Make sure the VM is stopped and was not scheduled for renaming already
+	vm, statusErr := app.fetchVirtualMachine(name, namespace)
+
+	if statusErr != nil {
+		writeError(statusErr, response)
+		return
+	}
+
+	// Check for renaming annotation on vm
+	if vm.Annotations != nil {
+		_, isScheduledForRenaming := vm.Annotations[v1.RenameToAnnotation]
+
+		if isScheduledForRenaming {
+			writeError(errors.NewConflict(v1.Resource("virtualmachine"),
+				name, fmt.Errorf("VM is already scheduled for renaming")), response)
+			return
+		}
+	}
+
+	// Make sure VM is stopped
+	if vm.Spec.Running == nil {
```

Comment:
Why not just use vm.RunStrategy()? Then there's no need for nested if's here.

Reply:
I was not aware of the existence of this method, just read it and I totally agree.
Changing it right now